src/main/java/org/ml_methods_group/algorithm/JMove.java

```diff
@@ -0,0 +1,276 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ml_methods_group.algorithm;
+
+
+import com.intellij.psi.*;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.ml_methods_group.algorithm.entity.ClassEntity;
+import org.ml_methods_group.algorithm.entity.MethodEntity;
+import org.ml_methods_group.config.Logging;
+
+import java.util.*;
+
+public class JMove extends Algorithm {
+    private static final Logger LOGGER = Logging.getLogger(JMove.class);
+    private final int MIN_NUMBER_OF_CANDIDATE_CLASSES = 3;
+    private final int MIN_NUMBER_OF_DEPENDENCIES = 4;
+    private final double MIN_DIFF_BETWEEN_SIMILARITY_COEFF_PERS = 0.25;
+
+    public JMove() {
+        super("JMove", false);
+    }
+
+    @Override
+    protected List<Refactoring> calculateRefactorings(ExecutionContext context, boolean enableFieldRefactorings) {
+        LOGGER.info("Starting calculating refactorings");
+        List<MethodEntity> allMethods = context.getEntities().getMethods();
+        List<ClassEntity> allClasses = context.getEntities().getClasses();
+        LOGGER.info("Found " + allMethods.size() + " methods and " + allClasses.size() + " classes");
+        List<Refactoring> refactorings = new ArrayList<>();
+
+        Map<String, ClassEntity> nameToClassEntity = new HashMap<>();
+        Map<String, Dependencies> nameToDependencies = new HashMap<>();
+
+        for(ClassEntity classEntity : allClasses) {
+            nameToClassEntity.put(classEntity.getName(), classEntity);
+        }
+
+        for(MethodEntity methodEntity : allMethods) {
+            nameToDependencies.put(methodEntity.getName(), new Dependencies(methodEntity, nameToClassEntity));
+        }
+
+
+        for(MethodEntity curMethod : allMethods) {
+            if(!curMethod.isMovable())
+                continue;
+            LOGGER.info("Checking " + curMethod.getName()); // todo some logging
+
+            Dependencies curDependencies = nameToDependencies.get(curMethod.getName());
+            ClassEntity curClass = nameToClassEntity.get(curMethod.getClassName());
+            if(curDependencies.cardinality() < MIN_NUMBER_OF_DEPENDENCIES //experimental
+                    || curClass.getRelevantProperties().getMethods().size() == 1 //because we calculate similarity between this method and all remaining in curClass
+                    || isGetter(curMethod) //this methods are rarely implemented in the wrong classes
+                    || isSetter(curMethod)) //todo: check if we need this check here and not in other place
+                continue;
+            double curSimilarity = calculateSimilarity(curMethod, curClass, nameToDependencies);
+            Map<ClassEntity, Double> potentialClasses = new HashMap<>();
```

Comment:
there is no need to store all potential target classes, you can simply introduce counter variable and use single potential class entity